### DISCUSSION TITLE
Recommended way of saving records when a record with a matching field already exists

### DISCUSSION THREAD

**Discussion by usrme** [pocketbase/pocketbase#2593] [Q&A] [answered]
I'm trying to load some data from a JSON and then just create records in a collection from those, which works fine, but what I'm after is that an overwrite should occur if a record with a matching field, say `name`, already exists. I could go down the route of checking prior to insertion whether a record with a matching `name` exists, delete the existing record beforehand, and then insert the one from the file, but I'm wondering if there's a better, more PocketBase-y way to go about this.

**Comment by ganigeorgiev** [comment]
There is no `UPSERT`-like support at the moment (_mainly because there is no easy way to distinguish the create and update events without a transaction lock_).

If you expect a lot of matching records it may be more efficient to eagerly fetch those records and perform an update with the new data. For example, if you are using the JS SDK, this could like:

```js
const json = [
    {name: "rec1", ...},
    {name: "rec2", ...},
];

// fetch all records with matching "name"
// note: if the json array is expected to be very big, it maybe also a good idea to batch the names filter (eg. per 1000)
const existing = await pb.collection("example").getFullList({
    // aka. `name = "rec1" || name = "rec2" || name = "rec3"`
    filter: json.map((item) => `name="${item.name}"`).join(" || ")
});

const promises = [];

for (let item in json) {
    let found = existing.find((exist) => item.name == exist.name);

    if (found) {
        promises.push(pb.collection("example").update(found.id, item, { $autoCancel: false }));
    } else {
        promises.push(pb.collection("example").create(item, { $autoCancel: false }));
    }
}

await Promise.all(promises);
```

**Reply by usrme** [reply]
My bad, I'm loading directly on the server, so I was wondering more specifically about Go.

**Reply by ganigeorgiev** [reply, marked as answer]
No, unfortunately the same apply when using Go. You can apply a similar approach by using something like:

```go
collection, err := app.Dao().FindCollectionByNameOrId("example")
if err != nil {
	return err
}

// remove the transaction if you don't need it
app.Dao().RunInTransaction(func(txDao *daos.Dao) error {
	for _, item := range json {
		record, _ := txDao.FindFirstRecordByData("example", "name", item["name"])
		if record == nil {
			record = models.NewRecord(collection)
		}

		record.Load(item)

		if err := txDao.SaveRecord(record); err != nil {
			return err
		}
	}

	return nil
})
```

**Reply by usrme** [reply]
Is the addition of the transaction just to ensure atomicity or is there anything else? :thinking:

**Reply by ganigeorgiev** [reply]
The transaction is there just to ensure atomicity, aka. either all records are inserted/updated or none (keep in mind that mixed read and write transaction as the above one will lock the db and other requests will have to wait the transaction to complete before proceeding).
